metagraph/src/graph/representation/canonical_dbg.cpp

```diff
@@ -577,5 +564,171 @@ void CanonicalDBG::reverse_complement(std::string &seq,
+    ::reverse_complement(rev_seq.begin(), rev_seq.end());
+    rev_seq.push_back('\0');
+
+    const auto &alphabet = graph_->alphabet();
+    for (size_t c = 0; c < alphabet.size(); ++c) {
+        if (alphabet[c] != boss::BOSS::kSentinel) {
+            rev_seq.back() = complement(alphabet[c]);
+            node_index prev = graph_->kmer_to_node(rev_seq);
+            if (prev != DeBruijnGraph::npos)
+                callback(prev);
+        }
+    }
+}
+
+void CanonicalDBG
+::adjacent_incoming_from_rc(node_index node,
+                            const std::function<void(node_index, edge_index)> &callback,
+                            const std::string &spelling_hint) const {
+    //        rshift    rc
+    // ATGGCT -> TGGCT* -> *AGCCA
+    if (const auto *dbg_succ_ = get_dbg_succ(*graph_)) {
+        //   AGAGGATCTCGTATGCCGTCTTCTGCTTGAG
+        //->  GAGGATCTCGTATGCCGTCTTCTGCTTGAG
+        //->  CTCAAGCAGAAGACGGCATACGAGATCCTC
+        const boss::BOSS &boss = dbg_succ_->get_boss();
+        auto *cache = get_extension_threadsafe<NodeFirstCache>();
+
+        bool is_dummy = (spelling_hint[0] == boss::BOSS::kSentinel);
+
+        boss::BOSS::edge_index rc_edge = (cache ? cache : fallback_cache_.get())->get_suffix_rc(
+            dbg_succ_->kmer_to_boss_index(node),
+            spelling_hint
+        );
+
+        if (!rc_edge)
+            return;
+
+        // rc_edge may be a dummy sink, so this won't work with graph_->call_incoming_kmers
+        boss.call_incoming_to_target(
+            cache ? cache->get_parent_pair(rc_edge).first : boss.bwd(rc_edge),
+            boss.get_node_last_value(rc_edge),
+            [&](boss::BOSS::edge_index incoming_boss_edge) {
+                node_index next = dbg_succ_->boss_to_kmer_index(incoming_boss_edge);
+                if (next == DeBruijnGraph::npos)
+                    return;
+
+                if (is_dummy) {
+                    char c = (cache ? cache : fallback_cache_.get())->get_first_char(
+                        incoming_boss_edge, rc_edge
+                    );
+                    if (c == boss::BOSS::kSentinel)
+                        return;
+                }
+
+                callback(next, rc_edge);
+            }
+        );
+
+        return;
+    }
+
+    // Do the checks by directly mapping the sequences of the desired k-mers.
+    // For non-DBGSuccinct graphs, this should be fast enough.
+    std::string rev_seq = graph_->get_node_sequence(node).substr(1) + std::string(1, '\0');
+    ::reverse_complement(rev_seq.begin(), rev_seq.end());
+    assert(rev_seq[0] == '\0');
+
+    const auto &alphabet = graph_->alphabet();
+    for (size_t c = 0; c < alphabet.size(); ++c) {
+        if (alphabet[c] != boss::BOSS::kSentinel) {
+            rev_seq[0] = complement(alphabet[c]);
+            node_index next = graph_->kmer_to_node(rev_seq);
+            if (next != DeBruijnGraph::npos)
+                callback(next, 0);
+        }
+    }
```

Comment:
Do in an `else` block. Otherwise it's easy to miss `return` above and assume that it's the ending of the function.
In general, try no to do `return` in super large `if` blocks like here.